packages/autocertifier-server/src/AutoCertifierServer.ts

```diff
@@ -0,0 +1,167 @@
+import { Database } from './Database'
+import { DnsServer } from './DnsServer'
+import { RestInterface } from './RestInterface'
+import { RestServer } from './RestServer'
+import { v4 } from 'uuid'
+import { CertifiedSubdomain, Session } from '@streamr/autocertifier-client'
+import { Logger } from '@streamr/utils'
+import { CertificateCreator } from './CertificateCreator'
+import { runStreamrChallenge } from './StreamrChallenger'
+import 'dotenv/config'
+import { ChallengeManager } from './ChallengeManager'
+
+const logger = new Logger(module)
+
+const validateEnvironmentVariable = (name: string): string | never => {
+    const value = process.env[name]
+    if (!value) {
+        throw new Error(`${name} environment variable is not set`)
+    }
+    return value
+}
+
+export class AutoCertifierServer implements RestInterface, ChallengeManager {
+
+    private domainName?: string
+    private dnsServer?: DnsServer
+    private restServer?: RestServer
+    private database?: Database
+    private certificateCreator?: CertificateCreator
+
+    // eslint-disable-next-line class-methods-use-this
+    public async createSession(): Promise<Session> {
+        logger.info('creating new session')
+        return { id: v4() }
+    }
+
+    public async createNewSubdomainAndCertificate(
+        ipAddress: string,
+        port: string,
+        streamrWebSocketPort: string,
+        sessionId: string
+    ): Promise<CertifiedSubdomain> {
+        logger.trace('Creating new subdomain and certificate for ' + ipAddress + ':' + port)
+
+        // this will throw if the client cannot answer the challenge of getting sessionId 
+        await runStreamrChallenge(ipAddress, streamrWebSocketPort, sessionId)
+        
+        const subdomain = v4()
+        const token = v4()
+        await this.database!.createSubdomain(subdomain, ipAddress, port, token)
+        const fqdn = subdomain + '.' + this.domainName
+        const certificate = await this.certificateCreator!.createCertificate(fqdn)
+        return { 
+            subdomain,
+            fqdn,
+            token,
+            certificate
+        }
+    }
+
+    public async createNewCertificateForSubdomain(
+        subdomain: string,
+        ipAddress: string,
+        port: string,
+        streamrWebSocketPort: string,
+        sessionId: string,
+        token: string
+    ): Promise<CertifiedSubdomain> {
+
+        logger.info('creating new certificate for ' + subdomain + ' and ' + ipAddress + ':' + port)
+
+        // This will throw if the token is incorrect
+        await this.updateSubdomainIp(subdomain, ipAddress, port, streamrWebSocketPort, sessionId, token)
+        const fqdn = subdomain + '.' + this.domainName
+        const certificate = await this.certificateCreator!.createCertificate(fqdn)
+        return {
+            subdomain,
```

Comment:
We don't need `subdomain` as the info is already in `fqdn`

(same pattern also elsewhere in the code base)